`get_call_trace` documentation (#1662)
foundry-rs/starknet-foundry: docs/src/appendix/snforge-library/get_call_trace.md

# `get_call_trace`

```rust
fn get_call_trace() -> CallTrace;
```

(For whole structure definition, please refer to [`snforge-std` source](https://github.com/foundry-rs/starknet-foundry/tree/v0.16.0/snforge_std))

> Gets latest call trace of the test, including the last call made to a contract.

Reviewer: I think this description is unclear. It should be sth like 
"Gets current call trace of the test up to the last call made to a contract."

## Example call trace
```cairo
use snforge_std::trace::{CallTrace, CallEntryPoint, CallType, EntryPointType, get_call_trace};
...

let ctrace = CallTrace {
    entry_point: CallEntryPoint {
        entry_point_type: EntryPointType::External,
        entry_point_selector: test_selector(),
        calldata: array![],
        contract_address: test_address(),
        caller_address: 0.try_into().unwrap(),
        call_type: CallType::Call,

>     },

Reviewer: We should mention that in a full trace this top entry point is fixed and symbolises call made to the test code (since we are in "invoke"), as it's not obvious

                        entry_point_type: EntryPointType::External,
                        entry_point_selector: selector!("other_entrypoint"),
                        calldata: array![1],
                        contract_address: 567.try_into().unwrap(),
                        caller_address: 123.try_into().unwrap(),
                        call_type: CallType::Call,
                    },
                    nested_calls: array![]
                }
            ],
        },
        CallTrace {
            entry_point: CallEntryPoint {
                entry_point_type: EntryPointType::L1Handler,
                entry_point_selector: selector!("handle_l1_message"),
                calldata: array![123, 567.into()],
                contract_address: 346.try_into().unwrap(),
                caller_address: 0.try_into().unwrap(),
                call_type: CallType::Call,
            },
            nested_calls: array![]
        }
    ]
};
```

The topmost-call is representing the test call, which will always be present if you're running a test.
It can have nested `CallTrace` - it's an array of subsequent traces made in scope of the call.

> The whole structure is represented as a tree of calls, in which each contract interaction (`constructor` call, `l1_handler` call, or a regular contract `call` via dispatcher) is a new execution scope - thus resulting in a new nested trace.

Reviewer: Forgot to mention library call via dispatcher

> 📝 **Note**
> 
> Failed (panicked) entrypoints will also register in the trace, 

> > if you handle the panic gracefully in the test. 

Reviewer: Missing be: "Failed (panicked) entrypoints will also BE register in the trace"

They will always be registered the trace, I don't think we need to mention that, it can be confusing. If the test panics bcs someone doesn't handle the panic it's obvious they won't be able to assert the trace later on.

# `get_call_trace`

```rust
fn get_call_trace() -> CallTrace;
```

> (For whole structure definition, please refer to [`snforge-std` source](https://github.com/foundry-rs/starknet-foundry/tree/00e126e54a8512c26ac0033b23d449d830b61195/snforge_std))

Reviewer 2: If we want to do it, tag would probably be better

```suggestion
(For whole structure definition, please refer to [`snforge-std` source](https://github.com/foundry-rs/starknet-foundry/tree/v0.16.0/snforge_std))
```

Gets latest call trace of the test, including the last call made to a contract.

## Example call trace

> ```

Reviewer 2: ```suggestion
```cairo
```

                        entry_point_type: EntryPointType::External,
                        entry_point_selector: selector!("other_entrypoint"),
                        calldata: array![1],
                        contract_address: 567.try_into().unwrap(),
                        caller_address: 123.try_into().unwrap(),
                        call_type: CallType::Call,
                    },
                    nested_calls: array![]
                }
            ],
        },
        CallTrace {
            entry_point: CallEntryPoint {
                entry_point_type: EntryPointType::L1Handler,
                entry_point_selector: selector!("handle_l1_message"),
                calldata: array![123, 567.into()],
                contract_address: 346.try_into().unwrap(),
                caller_address: 0.try_into().unwrap(),
                call_type: CallType::Call,
            },
            nested_calls: array![]
        }
    ]
};
```

The topmost-call is representing the test call, which will always be present if you're running a test.
It can have nested `CallTrace` - it's an array of subsequent traces made in scope of the call.

> The whole structure is represented as a tree of calls, in which each contract interaction (`constructor` call, `l1_handler` call, or a regular contract `call` via dispatcher) is a new execution scope - thus resulting
> in a new nested trace.

Reviewer 2: ```suggestion
The whole structure is represented as a tree of calls, in which each contract interaction (`constructor` call, `l1_handler` call, or a regular contract `call` via dispatcher) is a new execution scope - thus resulting in a new nested trace.
```

> 📝 **Note**
> 
> Failed (panicked) entrypoints will also register in the trace, 
> if you handle the panic gracefully in the test. 

## Displaying the trace

> The `CallTrace` structure implements also a `Display` trait, for a pretty-print with indentations:

Reviewer 2: ```suggestion
The `CallTrace` structure implements a `Display` trait, for a pretty-print with indentations:
```

> ```

Reviewer 2: ```suggestion
```cairo
```
